mgravell/ServiceStack.Redis
# Implement async APIs for ServiceStack.Redis

- Status
  - functionally complete
  - additional new tests compares the sync vs async interfaces
  - the interfaces presumably need to be carved back out to their owner projects
  - documentation?
- Approach
  - new interfaces have been added; where `IFoo` existed, there is now an `IFooAsync` twin
  - methods on `IFooAsync` have the `Async` suffix
  - all new methods are via **explicit** interface implementation - i.e. no `public` API changes, except:
    - `AsAsync()` helper API on client
    - `GetClientAsync()` (etc) extension methods on `IRedisClientsManager`
    - regular casts work (i.e. from `RedisClient` to `IRedisClientAsync`)
  - `ValueTask` preferred over `Task` (works best for pipelines/transactions, and is more flexible)
  - `CancellationToken` added throughout; this breaks `params` - will address via additional API methods
  - `ConfigureAwait(false)` used throughout library
  - where-ever possible, code is shared and reused between sync and async implementations; `.Async.cs` files and `partial class` used to append async functionality on suitable TFMs
  - a test ensures that all interfaces are common - i.e. for `IFoo`, it finds all types that implement `IFoo` and ensures that they also implement `IFooAsync`; one caveat here: `TrackingRedisClientsManager` is omitted (it won't play nice with `async` anyway, so it doesn't seem useful to do so)
- Testing
  - all relevant tests have been duplicated with `.Async.cs` files and marked with the `Async` trait; ~715 tests covering all areas of functionality
  - when `DEBUG` is defined, additional code activates that detects attempts to perform synchronous operations in an asynchronous context

Most migrations should be:

- add `AsAsync()`/cast/etc to the client/manager
- add `async`/`await` and `*Async` 

There are a few things that need to work differently - for example `Db` is now get-only, with the `SelectAsync` method supporting database switching
